src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -164,20 +171,87 @@ struct
                          applying the diff does not match blockchain state's \
                          ledger hash and staged ledger hash resp.\n")))
           in
-          {transition_with_hash; staged_ledger= transitioned_staged_ledger} )
+          { transition_with_hash
+          ; staged_ledger= transitioned_staged_ledger
+          ; ledger_proof_txns } )
 
     let hash {transition_with_hash; _} = With_hash.hash transition_with_hash
 
     let parent_hash {transition_with_hash; _} =
       Consensus.Protocol_state.previous_state_hash
-        (Inputs.External_transition.Verified.protocol_state
-           (With_hash.data transition_with_hash))
+        ( With_hash.data transition_with_hash
+        |> Inputs.External_transition.Verified.protocol_state )
+
+    (** Given:
+     *
+     *       bads
+     *        o       o
+     *       /       /
+     *    o ---- o -
+     *  root \  heir \
+     *        o       o
+     *               heir_heirs
+     *
+     *  commit heir into root and reparent heir_heirs onto root throw away bads
+    *)
+    let commit ~root ~heir ~bads ~heir_heirs :
+        t * t list * Transaction.t Non_empty_list.t option =
+      (* We shouldn't be merging breadcrumbs if the earlier one emitted a proof
+       * it needs to be committed into the snarked ledger *)
+      assert (Option.is_some root.ledger_proof_txns) ;
+      let f = Fn.compose Inputs.Staged_ledger.ledger staged_ledger in
+      let root_ledger = f root in
+      let heir_ledger = f heir in
+      let heir_heirs_ledgers = List.map ~f heir_heirs in
+      let bads_ledgers = List.map ~f bads in
+      (* tear out the heir's heirs from the heir *)
+      let dangling_masks =
+        List.map heir_heirs_ledgers ~f:(Ledger.unregister_mask_exn heir_ledger)
+      in
+      Ledger.Mask.Attached.commit heir_ledger ;
+      (* after we commit heir ledger, root ledger should reflect the changes from heir *)
+      assert (
+        Ledger_hash.equal
+          (Ledger.merkle_root heir_ledger)
+          (Ledger.merkle_root root_ledger) ) ;
```

Comment:
Since `commit` clears the mask it's called on, this doesn't actually check anything since the call to `merkl_root` is just forwarding to the `root_ledger` underlying the `heir_ledger`. You should instead compare against what's in the protocol state.